File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotLockOnSharedResource.cs
using System;
using System.Collections.Generic;

namespace Tests.Diagnostics
{
    public class LockOnThisOrType
    {
        private string myStringField;

        public void MyLockingMethod()
        {
            lock (this) // Noncompliant
//                ^^^^
            {
                // ...
            }

            lock (lockObj)
            {
                // ...
            }

            lock (typeof(LockOnThisOrType)) // Noncompliant
//                ^^^^^^^^^^^^^^^^^^^^^^^^
            {
                // ...
            }
            lock ((new LockOnThisOrType()).GetType()) // Noncompliant {{Lock on a dedicated object instance instead.}}
            {
                // ...
            }
            lock ("foo") // Noncompliant
            {
            }
            lock (myStringField) // Noncompliant
            {
            }
        }

        object lockObj = new object();
    }
}

File: analyzers/tests/SonarAnalyzer.Rules.Tests/Cases/DoNotLockOnSharedResource.vb
Imports System
Imports System.Collections.Generic

Namespace Tests.Diagnostics
    Public Class LockOnThisOrType
        Dim MyStringField As String

        Public Sub MyLockingMethod()
            SyncLock Me ' Noncompliant {{Lock on a dedicated object instance instead.}}
'                    ^^
            End SyncLock

            SyncLock lockObj
            End SyncLock

            SyncLock GetType(LockOnThisOrType) ' Noncompliant
'                    ^^^^^^^^^^^^^^^^^^^^^^^^^
            End SyncLock

            SyncLock (New LockOnThisOrType()).[GetType]() ' Noncompliant
'                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
            End SyncLock

            SyncLock ("foo") ' Noncompliant
            End SyncLock

            SyncLock (MyStringField) ' Noncompliant

            End SyncLock

        End Sub

        Dim lockObj As New Object()
    End Class
End Namespace
